rfc-157-edge-compute.md

```diff
@@ -0,0 +1,211 @@
+
+Origin request/response handlers allow us to do things that Compute@Edge does not yet support (but that are available in VCL), such as modifying a backend request after the cache key has been computed, or modifying a backend response before it gets cached.
+
+Origin request handlers can perform [dynamic origin selection](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-examples.html#lambda-examples-content-based-routing-examples), but unlike Compute@Edge, L@E provides no way to specify fallbacks in the case that one backend is down. We must instead use [CloudFront Origin Groups](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/high_availability_origin_failover.html#concept_origin_groups.lambda) for this purpose.
+
+## Choice of language
+
+TypeScript (or JavaScript if there's popular outcry against TS).
+
+Compute@Edge supports any [WASI](https://github.com/WebAssembly/WASI)-supporting language, and provides official SDKs for Rust, JavaScript, Go and AssemblyScript. The JavaScript SDK also includes [TypeScript definitions](https://github.com/fastly/js-compute-runtime/tree/main/types), and the default JavaScript project generated by the CLI includes Webpack (allowing us to use JavaScript modules).
+
+Lambda@Edge currently only supports two platforms: JavaScript (through Node.js), and Python. WASI-supporting languages like Rust _can_ be made to work on Lambda, but [it's pretty gnarly](https://markentier.tech/posts/2021/01/rust-wasm-on-aws-lambda-edge/) and maybe not the best idea. Lambda@Edge also has [(third-party) TypeScript definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/aws-lambda) available. Unlike Compute@Edge, there is no CLI for Lambda@Edge and no template projects; if we want to use JS modules we must bring our own JS bundler.
+
+As the only common language shared between the two platforms, JavaScript (or TypeScript) seems like the obvious choice.
+
+The primary benefit TypeScript provides over JavaScript is type safety, including enforcement of `readonly` properties. This will be particularly useful on the Lambda@Edge side because parts of the payload we receive from CloudFront are intended to be read-only. TypeScript enforces these invariants (potentially preventing us from introducing bugs), whereas JavaScript doesn't know how to.
+
+If folks have a strong aversion to TypeScript, JavaScript would also be a good option.
+
+## Architecture
+
+I propose that we build a three-layer architecture, with:
+
+- A platform-agnostic wrapper API to abstract away differences between the platform-specific types (e.g. `Request` and `Response`)
+- Platform-agnostic request and response handlers (corresponding to Lambda@Edge's viewer request and viewer response handlers concepts respectively), where the majority of the code lives, e.g.
```

Comment:
That's not quite right, I think I need to clarify a little. The C@E response handler might indeed contain logic, but it will be logic that corresponds to the viewer response handler on L@E (not the origin response handler).